File: 801-is-graph-bipartite/is-graph-bipartite.java
class Solution {

    private int[] color;
    private boolean bfs(int node, int[][] graph) 
    {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(node);
        color[node] = 1;
        while (!queue.isEmpty()) 
        {
            int currNode = queue.poll();
            for(int adjNode : graph[currNode])
            {
                if(color[adjNode] == -1)
                {
                    color[adjNode] = 1 - color[currNode];
                    queue.add(adjNode);
                }
                else if(color[adjNode] == color[currNode])  // Conflict found
                {
                    return false;  // Only return false on conflict
                }
            }
        }
        return true;
    }

    public boolean isBipartite(int[][] graph) 
    {
        int V = graph.length;
        color = new int[V];
        Arrays.fill(color,-1);

        for (int i = 0; i < V; i++) 
        {
            if (color[i] == -1) 
            {
                if(!bfs(i, graph))
                return false;
            }
        }
        return true;
    }
}

File: 801-is-graph-bipartite/is-graph-bipartite.cpp
class Solution {
public:
    bool dfs(int node, int col, vector<vector<int>>& graph, vector<bool>& vis, vector<int>& color)
    {
        if(vis[node])
        return color[node] == col;

        vis[node] = true;
        color[node] = col;
        
        for(int nbr : graph[node])
        {
            if(!dfs(nbr, 1 - col, graph, vis, color))
            return false;
        }
        
        return true;
    }

    bool isBipartite(vector<vector<int>>& graph) 
    {
        int n = graph.size();
        vector<bool> vis(n, false);  // Separate visited array
        vector<int> color(n, -1);    // Color array only for storing colors
        
        // Check each component of the graph
        for(int node = 0; node < n; node++)
        {
            if(!vis[node] && !dfs(node, 0, graph, vis, color))
            return false;
        }
        return true;
    }
};